drivers: Add Atmel SAM family I2C (TWI) driver

mnkp opened the pull request:
Added I2C bus (TWI) driver for Atmel SAM MCU family. Only
I2C Master Mode with 7 bit addressing is currently supported.

Origin: Original

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
drivers: Add Atmel SAM family I2C (TWI) driver

Added I2C bus (TWI) driver for Atmel SAM MCU family. Only
I2C Master Mode with 7 bit addressing is currently supported.

Tested on Arduino Due board.

Origin: Original

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

MaureenHelm committed:
dts: Add Kinetis SIM clock bindings

Adds device tree bindings for the Kinetis System Integration Module
(SIM), and defines peripheral source clocks (e.g., system clock or bus
clock) and clock gates for all Kinetis SoCs.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
clock_control: Introduce mcux sim driver

Adds a new clock control driver for Kinetis SoCs that have the system
integration (SIM) module. This will allow mcux shim drivers, such as
uart and i2c, to abstract the call to CLOCK_GetFreq() behind the
clock_control interface and thus be reused for SoCs with different clock
architectures.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
arm: Enable the mcux sim clock control driver on Kinetis SoCs

Because the mcux shim drivers will soon depend on a clock control
interface, enable the mcux sim clock control driver by default on all
Kinetis SoCs.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
serial: Refactor mcux uart shim driver to use clock control interface

Use the clock control interface instead of calling CLOCK_GetFreq()
directly. This will allow SoCs with different clock architectures to
reuse this driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
serial: Refactor mcux lpuart shim driver to use clock control interface

Refactors the mcux lpuart shim driver to use the clock control interface
instead of calling CLOCK_GetFreq() directly. This will allow SoCs with
different clock architectures to reuse this driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
serial: Refactor mcux lpsci shim driver to use clock control interface

Refactors the mcux lpsci shim driver to use the clock control interface
instead of calling CLOCK_GetFreq() directly. This will allow SoCs with
different clock architectures to reuse this driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

mnkp committed:
drivers: i2c: deprecate usage of i2c_atmel_sam3 driver

Atmel SAM3X series has been recently converted to use ASF and
should now use common SAM family drivers. The i2c_atmel_sam3
I2C driver will be removed in the future.

This commit also changes the default I2C driver for Arduino Due
board from i2c_atmel_sam3 to i2c_sam_twi.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

bryan-hunt commented:
Pinmap/pinmux should be moved to the board rather than reside under the SOC.

mnkp commented:
@bryan-hunt the pinmux code configures pins as defined by the Atmel SAM SoC family datasheets. It has nothing to do with the specifics of the board implementation/routing. Pins are configured only when the device instance is initialized, otherwise they are not touched.

bryan-hunt commented:
Yes it does match the SAM3/SAM4S series datasheets where the pinmux is many to one but for the SAMD or SAME series with more conventional pinmuxing the map is many to many so I was recommending the change for convention with the number of other boards already in the zephyr system (e.g. stm32f429i_disc1, arduino_101, frdm_k64f, etc). As I'm working on both the SAM4s Xplained Pro and SAMD21 Xplained Pro ports I'll leave the pinmap where it is in my pull request for the SAM4S and leave the question of convention up to the ARM arch guys.

mnkp commented:
Good point. If a signal can be routed to more than one pin on the SoC the choice should be made by the board config. This is handled currently by selecting a Kconfig option. E.g. to choose UART1 TXD pin on SAME70 SoC one has to set one of UART_SAM_PORT_1_PIN_TX_PA4, UART_SAM_PORT_1_PIN_TX_PA6, UART_SAM_PORT_1_PIN_TX_PD26 Kconfig options. This has to be ported to DTS. It's still on the TODO list.

By keeping the actual pin muxing code local to SoC we have only one copy of it.

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+	cfg.raw = config;
+
+	/* Configure clock */
+	switch (cfg.bits.speed) {
```
can we avoid using cfg.bits.speed and instead doing something like:

speed = (config & I2C_SPEED_MASK) >> I2C_SPEED_SHIFT;
switch (speed)

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+
+static void i2c0_sam_irq_config(void)
+{
+	IRQ_CONNECT(TWI0_IRQn, CONFIG_I2C_0_IRQ_PRI, i2c_sam_twi_isr,
```
we should get IRQ data from DTS

mnkp replied:
As I understood there is a plan to move to generate configuration data structures directly from DTS thus the requirement. That sounds reasonable. On the other hand the IRQ, peripheral ID, register addresses data are all available already from header files for all CMSIS compliant arm SoCs. So now we need to re-write this data in DTS format. That's extra work and it's error prone. Maybe there is another way to handle it? Is the config structure generation concept finalized and ready?

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+static const struct soc_gpio_pin pins_twi0[] = PINS_TWI0;
+
+static const struct i2c_sam_twi_dev_cfg i2c0_sam_config = {
+	.regs = TWI0,
```
we should get register address from DTS

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+
+static void i2c1_sam_irq_config(void)
+{
+	IRQ_CONNECT(TWI1_IRQn, CONFIG_I2C_1_IRQ_PRI, i2c_sam_twi_isr,
```
use DTS for IRQ info

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+static const struct soc_gpio_pin pins_twi1[] = PINS_TWI1;
+
+static const struct i2c_sam_twi_dev_cfg i2c1_sam_config = {
+	.regs = TWI1,
```
use DTS for reg addr.

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+	.regs = TWI0,
+	.irq_config = i2c0_sam_irq_config,
+	.periph_id = ID_TWI0,
+	.irq_id = TWI0_IRQn,
```
use DTS for IRQ info

galak commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,393 @@
+	.regs = TWI1,
+	.irq_config = i2c1_sam_irq_config,
+	.periph_id = ID_TWI1,
+	.irq_id = TWI1_IRQn,
```
use DTS for IRQ info

galak reviewed (changes requested):
some minor dts cleanups, and avoid use of bitfield for i2c cfg stuff

mnkp commented:
I have change usage of cfg.bits.speed as requested, also updated SAM4S Xplained board documentation adding support for I2C driver.

Could we handle the IRQ, peripheral ID, register addresses data from DTS issue when the config structure generation concept is ready? There are potentially more device drivers which will be affected.

dbkinder reviewed (dismissed):
+1 for doc changes

mnkp commented:
@galak I would like to rework twihs_sam I2C driver but it would create merge conflicts with this one if submitted simultaneously. Could we merge this driver if there are no further issues?

galak commented:
I would like to make the DT changes so its easier to tell what needs to get generated and see what patterns there are as we go down that path.

mnkp commented:
@galak the thing about IRQ number and register addresses is that we get them all for free from CMSIS header files. Going over DTS is a long way and is error prone; best avoided. However, if we want to generate device configuration data structure fully automatically then indeed DTS seems like the only way to go. The tricky part will be generating pin configuration data. I was asking about the concept because it's not clear to me what is the goal / what we want to achieve.

If we want to auto generate complete device configuration data structure then we also need to include Peripheral ID in DTS. Should I add it?

ydamigos committed:
ext: hal: stm32cube: Fix LSI_VALUE for F1, F3 families

This patch sets the correct LSI_VALUE, according to
STM32F3 reference manual's (RM0316) section 9.2.5 and
STM32F1 reference manual's (RM0008) section 7.2.5.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

galak commented:
Yes lets add Peripheral ID to the dts.  The intent is to get DTS to a point that SoC related data (address, irq, pins, clocks, etc) is coming from it, and we can use it to generate.  Which is why I ask to use the DT for the things we have, so we can start flushing out the things missing.

bryan-hunt commented:
I've yet to see an example for the SAM series where the peripheral id differs from the IRQ number since the peripheral id is used to enable the interrupt. Might I suggest just using the IRQ number unless there is a compelling reason (like an exception to the rule) to add it?

galak commented:
I think its better to introduce the peripheral id separate from the IRQ.  It saves us pain down the line when some SoC breaks the rule or have might have some peripheral with multiple IRQs.

mnkp committed:
dts: Add Peripheral ID field to Atmel SAM drivers

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
I've prepared requested update. I also added Peripheral ID to the configurations of all the remaining Atmel SAM DTS drivers (as a separate commit).

galak commented on dts/arm/atmel/sam3x.dtsi:
```diff
@@ -49,14 +51,16 @@
 			reg = <0x400A0000 0x130>;
-			interrupts = <18 0>;
+			interrupts = <19 0>;
+			pid = <19>;
```
do you mind either putting the fix for uart in a separate commit, or at least comment in the commit message about fixing this.

mnkp replied:
Sorry about that, I put it now in a separate commit. s/pid/peripheral-id/g is done too.

galak commented on dts/arm/atmel/sam3x.dtsi:
```diff
@@ -33,6 +33,7 @@
 			compatible = "atmel,sam-uart";
 			reg = <0x400E0800 0x124>;
 			interrupts = <8 0>;
+			pid = <8>;
```
can we use peripheral-id instead of pid?

galak reviewed (commented):
1. how about using 'peripheral-id'.  (pid is too generic and can mean a number of things).
2. minor nit about fix of uart irq, and commenting it or splitting the fix

mnkp committed:
dts: sam3x: fix incorrect usart2 IRQ number

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

MaureenHelm commented on boards/arm/arduino_due/Kconfig.defconfig:
```diff
@@ -40,20 +40,15 @@ endif # GPIO
 if I2C
 
-config I2C_ATMEL_SAM3
+config I2C_SAM_TWI
```
Please add a comment in the commit message to note that you're changing the default I2C driver for this board

mnkp replied:
Done.

MaureenHelm commented on drivers/i2c/i2c_sam_twi.c:
```diff
@@ -0,0 +1,390 @@
+
+	/* Not Acknowledged */
+	if (isr_status & TWI_SR_NACK) {
+		msg->twi_sr = isr_status;
```
Do you need to give the semaphore here?

mnkp replied:
According to the datasheet TXCOMP is always set alongside NACK but the code indeed could have been written better. I updated it to make it more explicit.
